File: Stage25/Java/Problem2110.java
package Stage25.Java;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Problem2110 {
    static int N, C;
    static int[] house;

    static int binarySearch() {
        int left = 1;
        int right = house[N - 1] - house[0];
        int answer = 0;

        while (left <= right) {
            int mid = (left + right) / 2;

            if (canInstall(mid)) {
                answer = mid;
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        return answer;
    }

    static boolean canInstall(int dist) {
        int count = 1;
        int last = house[0];

        for (int i = 1; i < N; i++) {
            if (house[i] - last >= dist) {
                count++;
                last = house[i];
            }
        }

        return count >= C;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        C = Integer.parseInt(st.nextToken());

        house = new int[N];
        for (int i = 0; i < N; i++) {
            house[i] = Integer.parseInt(br.readLine());
        }

        Arrays.sort(house);
        System.out.println(binarySearch());
    }
}

File: Stage25/Python/Problem2110.py
import sys
input = sys.stdin.readline

def can_install(dist, house, N, C):
    count = 1
    last = house[0]

    for i in range(1, N):
        if house[i] - last >= dist:
            count += 1
            last = house[i]

    return count >= C


def binary_search(house, N, C):
    left = 1
    right = house[-1] - house[0]
    answer = 0

    while left <= right:
        mid = (left + right) // 2

        if can_install(mid, house, N, C):
            answer = mid
            left = mid + 1
        else:
            right = mid - 1

    return answer


def main():
    N, C = map(int, input().split())
    house = [int(input()) for _ in range(N)]
    house.sort()

    print(binary_search(house, N, C))


if __name__ == "__main__":
    main()